// @mcbe/ruacore_server/index.ts

import { createBroadcast, evitem, createListener } from "@mcbe/event";
import Event from "krevent";
import { system } from "@mcbe/system_server";

const ruacore = {
    initialize:new Event<()=>void>(),
    shutdown:new Event<()=>void>(),
    update:new Event<()=>void>(),
    broadcast:createBroadcast({
        DisplayChat: evitem<IDisplayChatParameters>(SendToMinecraftServer.DisplayChat),
        ExecuteCommand: evitem<IExecuteCommandParameters>(SendToMinecraftServer.ExecuteCommand),
        PlaySound: evitem<IPlaySoundParameters>(SendToMinecraftServer.PlaySound),
        ScriptLoggerConfig: evitem<IScriptLoggerConfigParameters>(SendToMinecraftServer.ScriptLoggerConfig),
        SpawnParticleAttachedEntity: evitem<ISpawnParticleAttachedEntityParameters>(SendToMinecraftServer.SpawnParticleAttachedEntity),
        SpawnParticleInWorld: evitem<ISpawnParticleInWorldParameters>(SendToMinecraftServer.SpawnParticleInWorld),
    }),    
    listen:createListener({
        BlockDestructionStarted: evitem<IBlockDestructionStartedEventData>(ReceiveFromMinecraftServer.BlockDestructionStarted),
        BlockDestructionStopped: evitem<IBlockDestructionStoppedEventData>(ReceiveFromMinecraftServer.BlockDestructionStopped),
        BlockInteractedWith: evitem<IBlockInteractedWithEventData>(ReceiveFromMinecraftServer.BlockInteractedWith),
        EntityAcquiredItem: evitem<IEntityAcquiredItemEventData>(ReceiveFromMinecraftServer.EntityAcquiredItem),
        EntityCarriedItemChanged: evitem<IEntityCarriedItemChangedEventData>(ReceiveFromMinecraftServer.EntityCarriedItemChanged),
        EntityCreated: evitem<IEntityCreatedEventData>(ReceiveFromMinecraftServer.EntityCreated),
        EntityDeath: evitem<IEntityDeathEventData>(ReceiveFromMinecraftServer.EntityDeath),
        EntityDroppedItem: evitem<IEntityDroppedItemEventData>(ReceiveFromMinecraftServer.EntityDroppedItem),
        EntityEquippedArmor: evitem<IEntityEquippedArmorEventData>(ReceiveFromMinecraftServer.EntityEquippedArmor),
        EntityStartRiding: evitem<IEntityStartRidingEventData>(ReceiveFromMinecraftServer.EntityStartRiding),
        EntityStopRiding: evitem<IEntityStopRidingEventData>(ReceiveFromMinecraftServer.EntityStopRiding),
        EntityTick: evitem<IEntityTickEventData>(ReceiveFromMinecraftServer.EntityTick),
        EntityUseItem: evitem<IEntityUseItemEventData>(ReceiveFromMinecraftServer.EntityUseItem),
        PistonMovedBlock: evitem<IPistonMovedBlockEventData>(ReceiveFromMinecraftServer.PistonMovedBlock),
        PlaySound: evitem<IPlaySoundEventData>(ReceiveFromMinecraftServer.PlaySound),
        PlayerAttackedEntity: evitem<IPlayerAttackedEntityEventData>(ReceiveFromMinecraftServer.PlayerAttackedEntity),
        PlayerDestroyedBlock: evitem<IPlayerDestroyedBlockEventData>(ReceiveFromMinecraftServer.PlayerDestroyedBlock),
        PlayerPlacedBlock: evitem<IPlayerPlacedBlockEventData>(ReceiveFromMinecraftServer.PlayerPlacedBlock),
        WeatherChangekey: evitem<IWeatherChangedEventData>(ReceiveFromMinecraftServer.WeatherChanged)
    }),
};
Object.freeze(ruacore);

system.initialize = ()=>ruacore.initialize.fire();
system.shutdown = ()=>ruacore.shutdown.fire();
system.update = ()=>ruacore.update.fire();
Object.freeze(system);

export default ruacore;
