// @mcbe/identifier/id.ts

import MAKEID from "./make";

const ID = {
    acacia_button: MAKEID.acacia_button,
    acacia_door: MAKEID.acacia_door,
    acacia_fence_gate: MAKEID.acacia_fence_gate,
    acacia_pressure_plate: MAKEID.acacia_pressure_plate,
    acacia_sign: MAKEID.acacia_sign,
    acacia_stairs: MAKEID.acacia_stairs,
    acacia_standing_sign: MAKEID.acacia_standing_sign,
    acacia_trapdoor: MAKEID.acacia_trapdoor,
    acacia_wall_sign: MAKEID.acacia_wall_sign,
    activator_rail: MAKEID.activator_rail,
    air: MAKEID.air,
    andesite_stairs: MAKEID.andesite_stairs,
    anvil: MAKEID.anvil,
    apple: MAKEID.apple,
    appleenchanted: MAKEID.appleenchanted,
    area_effect_cloud: MAKEID.area_effect_cloud,
    armor_stand: MAKEID.armor_stand,
    arrow: MAKEID.arrow,
    baked_potato: MAKEID.baked_potato,
    bamboo: MAKEID.bamboo,
    bamboo_sapling: MAKEID.bamboo_sapling,
    banner: MAKEID.banner,
    banner_pattern: MAKEID.banner_pattern,
    barrel: MAKEID.barrel,
    barrier: MAKEID.barrier,
    bat: MAKEID.bat,
    beacon: MAKEID.beacon,
    bed: MAKEID.bed,
    bedrock: MAKEID.bedrock,
    bee: MAKEID.bee,
    beef: MAKEID.beef,
    beehive: MAKEID.beehive,
    beetroot: MAKEID.beetroot,
    beetroot_seeds: MAKEID.beetroot_seeds,
    beetroot_soup: MAKEID.beetroot_soup,
    bell: MAKEID.bell,
    birch_button: MAKEID.birch_button,
    birch_door: MAKEID.birch_door,
    birch_fence_gate: MAKEID.birch_fence_gate,
    birch_pressure_plate: MAKEID.birch_pressure_plate,
    birch_sign: MAKEID.birch_sign,
    birch_stairs: MAKEID.birch_stairs,
    birch_standing_sign: MAKEID.birch_standing_sign,
    birch_trapdoor: MAKEID.birch_trapdoor,
    birch_wall_sign: MAKEID.birch_wall_sign,
    black_glazed_terracotta: MAKEID.black_glazed_terracotta,
    blast_furnace: MAKEID.blast_furnace,
    blaze: MAKEID.blaze,
    blaze_powder: MAKEID.blaze_powder,
    blaze_rod: MAKEID.blaze_rod,
    blue_glazed_terracotta: MAKEID.blue_glazed_terracotta,
    blue_ice: MAKEID.blue_ice,
    boat: MAKEID.boat,
    bone: MAKEID.bone,
    bone_block: MAKEID.bone_block,
    book: MAKEID.book,
    bookshelf: MAKEID.bookshelf,
    bow: MAKEID.bow,
    bowl: MAKEID.bowl,
    bread: MAKEID.bread,
    brewing_stand: MAKEID.brewing_stand,
    brick: MAKEID.brick,
    brick_block: MAKEID.brick_block,
    brick_stairs: MAKEID.brick_stairs,
    brown_glazed_terracotta: MAKEID.brown_glazed_terracotta,
    brown_mushroom: MAKEID.brown_mushroom,
    brown_mushroom_block: MAKEID.brown_mushroom_block,
    bucket: MAKEID.bucket,
    cactus: MAKEID.cactus,
    cake: MAKEID.cake,
    camera: MAKEID.camera,
    campfire: MAKEID.campfire,
    carpet: MAKEID.carpet,
    carrot: MAKEID.carrot,
    carrotonastick: MAKEID.carrotonastick,
    carrots: MAKEID.carrots,
    cartography_table: MAKEID.cartography_table,
    carved_pumpkin: MAKEID.carved_pumpkin,
    cat: MAKEID.cat,
    cauldron: MAKEID.cauldron,
    chainmail_boots: MAKEID.chainmail_boots,
    chainmail_chestplate: MAKEID.chainmail_chestplate,
    chainmail_helmet: MAKEID.chainmail_helmet,
    chainmail_leggings: MAKEID.chainmail_leggings,
    chest: MAKEID.chest,
    chest_minecart: MAKEID.chest_minecart,
    chicken: MAKEID.chicken,
    chorus_fruit: MAKEID.chorus_fruit,
    chorus_fruit_popped: MAKEID.chorus_fruit_popped,
    clay: MAKEID.clay,
    clay_ball: MAKEID.clay_ball,
    clock: MAKEID.clock,
    clownfish: MAKEID.clownfish,
    coal: MAKEID.coal,
    coal_block: MAKEID.coal_block,
    coal_ore: MAKEID.coal_ore,
    cobblestone: MAKEID.cobblestone,
    cobblestone_wall: MAKEID.cobblestone_wall,
    cocoa: MAKEID.cocoa,
    cod: MAKEID.cod,
    command_block: MAKEID.command_block,
    comparator: MAKEID.comparator,
    compass: MAKEID.compass,
    composter: MAKEID.composter,
    concrete: MAKEID.concrete,
    concrete_powder: MAKEID.concrete_powder,
    conduit: MAKEID.conduit,
    cooked_beef: MAKEID.cooked_beef,
    cooked_chicken: MAKEID.cooked_chicken,
    cooked_fish: MAKEID.cooked_fish,
    cooked_porkchop: MAKEID.cooked_porkchop,
    cooked_rabbit: MAKEID.cooked_rabbit,
    cooked_salmon: MAKEID.cooked_salmon,
    cookie: MAKEID.cookie,
    cow: MAKEID.cow,
    crafting_table: MAKEID.crafting_table,
    creeper: MAKEID.creeper,
    crossbow: MAKEID.crossbow,
    cyan_glazed_terracotta: MAKEID.cyan_glazed_terracotta,
    dark_oak_button: MAKEID.dark_oak_button,
    dark_oak_door: MAKEID.dark_oak_door,
    dark_oak_fence_gate: MAKEID.dark_oak_fence_gate,
    dark_oak_pressure_plate: MAKEID.dark_oak_pressure_plate,
    dark_oak_stairs: MAKEID.dark_oak_stairs,
    dark_oak_trapdoor: MAKEID.dark_oak_trapdoor,
    dark_prismarine_stairs: MAKEID.dark_prismarine_stairs,
    darkoak_sign: MAKEID.darkoak_sign,
    darkoak_standing_sign: MAKEID.darkoak_standing_sign,
    darkoak_wall_sign: MAKEID.darkoak_wall_sign,
    daylight_detector: MAKEID.daylight_detector,
    daylight_detector_inverted: MAKEID.daylight_detector_inverted,
    deadbush: MAKEID.deadbush,
    detector_rail: MAKEID.detector_rail,
    diamond: MAKEID.diamond,
    diamond_axe: MAKEID.diamond_axe,
    diamond_block: MAKEID.diamond_block,
    diamond_boots: MAKEID.diamond_boots,
    diamond_chestplate: MAKEID.diamond_chestplate,
    diamond_helmet: MAKEID.diamond_helmet,
    diamond_hoe: MAKEID.diamond_hoe,
    diamond_leggings: MAKEID.diamond_leggings,
    diamond_ore: MAKEID.diamond_ore,
    diamond_pickaxe: MAKEID.diamond_pickaxe,
    diamond_shovel: MAKEID.diamond_shovel,
    diamond_sword: MAKEID.diamond_sword,
    diorite_stairs: MAKEID.diorite_stairs,
    dirt: MAKEID.dirt,
    dispenser: MAKEID.dispenser,
    dolphin: MAKEID.dolphin,
    donkey: MAKEID.donkey,
    double_plant: MAKEID.double_plant,
    double_stone_slab: MAKEID.double_stone_slab,
    double_stone_slab2: MAKEID.double_stone_slab2,
    double_stone_slab3: MAKEID.double_stone_slab3,
    double_stone_slab4: MAKEID.double_stone_slab4,
    double_wooden_slab: MAKEID.double_wooden_slab,
    dragon_breath: MAKEID.dragon_breath,
    dragon_egg: MAKEID.dragon_egg,
    dried_kelp: MAKEID.dried_kelp,
    dried_kelp_block: MAKEID.dried_kelp_block,
    dropper: MAKEID.dropper,
    drowned: MAKEID.drowned,
    dye: MAKEID.dye,
    egg: MAKEID.egg,
    element_0: MAKEID.element_0,
    element_1: MAKEID.element_1,
    element_10: MAKEID.element_10,
    element_100: MAKEID.element_100,
    element_101: MAKEID.element_101,
    element_102: MAKEID.element_102,
    element_103: MAKEID.element_103,
    element_104: MAKEID.element_104,
    element_105: MAKEID.element_105,
    element_106: MAKEID.element_106,
    element_107: MAKEID.element_107,
    element_108: MAKEID.element_108,
    element_109: MAKEID.element_109,
    element_11: MAKEID.element_11,
    element_110: MAKEID.element_110,
    element_111: MAKEID.element_111,
    element_112: MAKEID.element_112,
    element_113: MAKEID.element_113,
    element_114: MAKEID.element_114,
    element_115: MAKEID.element_115,
    element_116: MAKEID.element_116,
    element_117: MAKEID.element_117,
    element_118: MAKEID.element_118,
    element_12: MAKEID.element_12,
    element_13: MAKEID.element_13,
    element_14: MAKEID.element_14,
    element_15: MAKEID.element_15,
    element_16: MAKEID.element_16,
    element_17: MAKEID.element_17,
    element_18: MAKEID.element_18,
    element_19: MAKEID.element_19,
    element_2: MAKEID.element_2,
    element_20: MAKEID.element_20,
    element_21: MAKEID.element_21,
    element_22: MAKEID.element_22,
    element_23: MAKEID.element_23,
    element_24: MAKEID.element_24,
    element_25: MAKEID.element_25,
    element_26: MAKEID.element_26,
    element_27: MAKEID.element_27,
    element_28: MAKEID.element_28,
    element_29: MAKEID.element_29,
    element_3: MAKEID.element_3,
    element_30: MAKEID.element_30,
    element_31: MAKEID.element_31,
    element_32: MAKEID.element_32,
    element_33: MAKEID.element_33,
    element_34: MAKEID.element_34,
    element_35: MAKEID.element_35,
    element_36: MAKEID.element_36,
    element_37: MAKEID.element_37,
    element_38: MAKEID.element_38,
    element_39: MAKEID.element_39,
    element_4: MAKEID.element_4,
    element_40: MAKEID.element_40,
    element_41: MAKEID.element_41,
    element_42: MAKEID.element_42,
    element_43: MAKEID.element_43,
    element_44: MAKEID.element_44,
    element_45: MAKEID.element_45,
    element_46: MAKEID.element_46,
    element_47: MAKEID.element_47,
    element_48: MAKEID.element_48,
    element_49: MAKEID.element_49,
    element_5: MAKEID.element_5,
    element_50: MAKEID.element_50,
    element_51: MAKEID.element_51,
    element_52: MAKEID.element_52,
    element_53: MAKEID.element_53,
    element_54: MAKEID.element_54,
    element_55: MAKEID.element_55,
    element_56: MAKEID.element_56,
    element_57: MAKEID.element_57,
    element_58: MAKEID.element_58,
    element_59: MAKEID.element_59,
    element_6: MAKEID.element_6,
    element_60: MAKEID.element_60,
    element_61: MAKEID.element_61,
    element_62: MAKEID.element_62,
    element_63: MAKEID.element_63,
    element_64: MAKEID.element_64,
    element_65: MAKEID.element_65,
    element_66: MAKEID.element_66,
    element_67: MAKEID.element_67,
    element_68: MAKEID.element_68,
    element_69: MAKEID.element_69,
    element_7: MAKEID.element_7,
    element_70: MAKEID.element_70,
    element_71: MAKEID.element_71,
    element_72: MAKEID.element_72,
    element_73: MAKEID.element_73,
    element_74: MAKEID.element_74,
    element_75: MAKEID.element_75,
    element_76: MAKEID.element_76,
    element_77: MAKEID.element_77,
    element_78: MAKEID.element_78,
    element_79: MAKEID.element_79,
    element_8: MAKEID.element_8,
    element_80: MAKEID.element_80,
    element_81: MAKEID.element_81,
    element_82: MAKEID.element_82,
    element_83: MAKEID.element_83,
    element_84: MAKEID.element_84,
    element_85: MAKEID.element_85,
    element_86: MAKEID.element_86,
    element_87: MAKEID.element_87,
    element_88: MAKEID.element_88,
    element_89: MAKEID.element_89,
    element_9: MAKEID.element_9,
    element_90: MAKEID.element_90,
    element_91: MAKEID.element_91,
    element_92: MAKEID.element_92,
    element_93: MAKEID.element_93,
    element_94: MAKEID.element_94,
    element_95: MAKEID.element_95,
    element_96: MAKEID.element_96,
    element_97: MAKEID.element_97,
    element_98: MAKEID.element_98,
    element_99: MAKEID.element_99,
    elytra: MAKEID.elytra,
    emerald: MAKEID.emerald,
    emerald_block: MAKEID.emerald_block,
    emerald_ore: MAKEID.emerald_ore,
    emptymap: MAKEID.emptymap,
    enchanted_book: MAKEID.enchanted_book,
    enchanting_table: MAKEID.enchanting_table,
    end_brick_stairs: MAKEID.end_brick_stairs,
    end_bricks: MAKEID.end_bricks,
    end_crystal: MAKEID.end_crystal,
    end_portal: MAKEID.end_portal,
    end_portal_frame: MAKEID.end_portal_frame,
    end_rod: MAKEID.end_rod,
    end_stone: MAKEID.end_stone,
    ender_chest: MAKEID.ender_chest,
    ender_eye: MAKEID.ender_eye,
    ender_pearl: MAKEID.ender_pearl,
    enderman: MAKEID.enderman,
    evocation_illager: MAKEID.evocation_illager,
    experience_bottle: MAKEID.experience_bottle,
    eye_of_ender_signal: MAKEID.eye_of_ender_signal,
    farmland: MAKEID.farmland,
    feather: MAKEID.feather,
    fence: MAKEID.fence,
    fence_gate: MAKEID.fence_gate,
    fermented_spider_eye: MAKEID.fermented_spider_eye,
    fire: MAKEID.fire,
    fireball: MAKEID.fireball,
    fireworks: MAKEID.fireworks,
    fireworkscharge: MAKEID.fireworkscharge,
    fish: MAKEID.fish,
    fishing_hook: MAKEID.fishing_hook,
    fishing_rod: MAKEID.fishing_rod,
    fletching_table: MAKEID.fletching_table,
    flint: MAKEID.flint,
    flint_and_steel: MAKEID.flint_and_steel,
    flower_pot: MAKEID.flower_pot,
    flowing_lava: MAKEID.flowing_lava,
    flowing_water: MAKEID.flowing_water,
    fox: MAKEID.fox,
    frame: MAKEID.frame,
    frosted_ice: MAKEID.frosted_ice,
    furnace: MAKEID.furnace,
    gateway: MAKEID.gateway,
    ghast: MAKEID.ghast,
    ghast_tear: MAKEID.ghast_tear,
    glass: MAKEID.glass,
    glass_bottle: MAKEID.glass_bottle,
    glass_pane: MAKEID.glass_pane,
    glowstone: MAKEID.glowstone,
    glowstone_dust: MAKEID.glowstone_dust,
    gold_block: MAKEID.gold_block,
    gold_ingot: MAKEID.gold_ingot,
    gold_nugget: MAKEID.gold_nugget,
    gold_ore: MAKEID.gold_ore,
    golden_apple: MAKEID.golden_apple,
    golden_axe: MAKEID.golden_axe,
    golden_boots: MAKEID.golden_boots,
    golden_carrot: MAKEID.golden_carrot,
    golden_chestplate: MAKEID.golden_chestplate,
    golden_helmet: MAKEID.golden_helmet,
    golden_hoe: MAKEID.golden_hoe,
    golden_leggings: MAKEID.golden_leggings,
    golden_pickaxe: MAKEID.golden_pickaxe,
    golden_rail: MAKEID.golden_rail,
    golden_shovel: MAKEID.golden_shovel,
    golden_sword: MAKEID.golden_sword,
    granite_stairs: MAKEID.granite_stairs,
    grass: MAKEID.grass,
    grass_pass: MAKEID.grass_pass,
    gravel: MAKEID.gravel,
    gray_glazed_terracotta: MAKEID.gray_glazed_terracotta,
    green_glazed_terracotta: MAKEID.green_glazed_terracotta,
    grindstone: MAKEID.grindstone,
    guardian: MAKEID.guardian,
    gunpowder: MAKEID.gunpowder,
    hardened_clay: MAKEID.hardened_clay,
    hay_block: MAKEID.hay_block,
    heart_of_the_sea: MAKEID.heart_of_the_sea,
    heavy_weighted_pressure_plate: MAKEID.heavy_weighted_pressure_plate,
    hoglin: MAKEID.hoglin,
    honey_block: MAKEID.honey_block,
    honey_bottle: MAKEID.honey_bottle,
    honeycomb: MAKEID.honeycomb,
    honeycomb_block: MAKEID.honeycomb_block,
    hopper: MAKEID.hopper,
    hopper_minecart: MAKEID.hopper_minecart,
    horse: MAKEID.horse,
    horsearmorgold: MAKEID.horsearmorgold,
    horsearmoriron: MAKEID.horsearmoriron,
    horsearmorleather: MAKEID.horsearmorleather,
    husk: MAKEID.husk,
    ice: MAKEID.ice,
    iron_axe: MAKEID.iron_axe,
    iron_bars: MAKEID.iron_bars,
    iron_block: MAKEID.iron_block,
    iron_boots: MAKEID.iron_boots,
    iron_chestplate: MAKEID.iron_chestplate,
    iron_door: MAKEID.iron_door,
    iron_helmet: MAKEID.iron_helmet,
    iron_hoe: MAKEID.iron_hoe,
    iron_ingot: MAKEID.iron_ingot,
    iron_leggings: MAKEID.iron_leggings,
    iron_nugget: MAKEID.iron_nugget,
    iron_ore: MAKEID.iron_ore,
    iron_pickaxe: MAKEID.iron_pickaxe,
    iron_shovel: MAKEID.iron_shovel,
    iron_sword: MAKEID.iron_sword,
    iron_trapdoor: MAKEID.iron_trapdoor,
    item: MAKEID.item,
    itemframe: MAKEID.itemframe,
    jukebox: MAKEID.jukebox,
    jungle_button: MAKEID.jungle_button,
    jungle_door: MAKEID.jungle_door,
    jungle_fence_gate: MAKEID.jungle_fence_gate,
    jungle_pressure_plate: MAKEID.jungle_pressure_plate,
    jungle_sign: MAKEID.jungle_sign,
    jungle_stairs: MAKEID.jungle_stairs,
    jungle_standing_sign: MAKEID.jungle_standing_sign,
    jungle_trapdoor: MAKEID.jungle_trapdoor,
    jungle_wall_sign: MAKEID.jungle_wall_sign,
    kelp: MAKEID.kelp,
    ladder: MAKEID.ladder,
    lantern: MAKEID.lantern,
    lapis_block: MAKEID.lapis_block,
    lapis_ore: MAKEID.lapis_ore,
    lava: MAKEID.lava,
    lead: MAKEID.lead,
    leather: MAKEID.leather,
    leather_boots: MAKEID.leather_boots,
    leather_chestplate: MAKEID.leather_chestplate,
    leather_helmet: MAKEID.leather_helmet,
    leather_leggings: MAKEID.leather_leggings,
    leaves: MAKEID.leaves,
    leaves2: MAKEID.leaves2,
    lectern: MAKEID.lectern,
    lever: MAKEID.lever,
    light_block: MAKEID.light_block,
    light_blue_glazed_terracotta: MAKEID.light_blue_glazed_terracotta,
    light_weighted_pressure_plate: MAKEID.light_weighted_pressure_plate,
    lime_glazed_terracotta: MAKEID.lime_glazed_terracotta,
    lingering_potion: MAKEID.lingering_potion,
    lit_blast_furnace: MAKEID.lit_blast_furnace,
    lit_furnace: MAKEID.lit_furnace,
    lit_pumpkin: MAKEID.lit_pumpkin,
    lit_smoker: MAKEID.lit_smoker,
    llama: MAKEID.llama,
    log: MAKEID.log,
    log2: MAKEID.log2,
    loom: MAKEID.loom,
    magenta_glazed_terracotta: MAKEID.magenta_glazed_terracotta,
    magma: MAKEID.magma,
    magma_cream: MAKEID.magma_cream,
    magma_cube: MAKEID.magma_cube,
    map: MAKEID.map,
    melon: MAKEID.melon,
    melon_block: MAKEID.melon_block,
    melon_seeds: MAKEID.melon_seeds,
    melon_stem: MAKEID.melon_stem,
    minecart: MAKEID.minecart,
    mob_spawner: MAKEID.mob_spawner,
    monster_egg: MAKEID.monster_egg,
    mooshroom: MAKEID.mooshroom,
    mossy_cobblestone: MAKEID.mossy_cobblestone,
    mossy_cobblestone_stairs: MAKEID.mossy_cobblestone_stairs,
    mossy_stone_brick_stairs: MAKEID.mossy_stone_brick_stairs,
    mushroom_stew: MAKEID.mushroom_stew,
    muttoncooked: MAKEID.muttoncooked,
    muttonraw: MAKEID.muttonraw,
    mycelium: MAKEID.mycelium,
    name_tag: MAKEID.name_tag,
    nautilus_shell: MAKEID.nautilus_shell,
    nether_brick: MAKEID.nether_brick,
    nether_brick_fence: MAKEID.nether_brick_fence,
    nether_brick_stairs: MAKEID.nether_brick_stairs,
    nether_wart: MAKEID.nether_wart,
    nether_wart_block: MAKEID.nether_wart_block,
    netherbrick: MAKEID.netherbrick,
    netherrack: MAKEID.netherrack,
    netherstar: MAKEID.netherstar,
    normal_stone_stairs: MAKEID.normal_stone_stairs,
    noteblock: MAKEID.noteblock,
    oak_pressure_plate: MAKEID.oak_pressure_plate,
    oak_stairs: MAKEID.oak_stairs,
    observer: MAKEID.observer,
    obsidian: MAKEID.obsidian,
    ocelot: MAKEID.ocelot,
    orange_glazed_terracotta: MAKEID.orange_glazed_terracotta,
    packed_ice: MAKEID.packed_ice,
    painting: MAKEID.painting,
    panda: MAKEID.panda,
    paper: MAKEID.paper,
    parrot: MAKEID.parrot,
    phantom: MAKEID.phantom,
    phantom_membrane: MAKEID.phantom_membrane,
    pig: MAKEID.pig,
    piglin: MAKEID.piglin,
    pillager: MAKEID.pillager,
    pillager_patrol: MAKEID.pillager_patrol,
    pink_glazed_terracotta: MAKEID.pink_glazed_terracotta,
    piston: MAKEID.piston,
    pistonArmCollision: MAKEID.pistonArmCollision,
    planks: MAKEID.planks,
    player: MAKEID.player,
    poisonous_potato: MAKEID.poisonous_potato,
    polar_bear: MAKEID.polar_bear,
    polished_andesite_stairs: MAKEID.polished_andesite_stairs,
    polished_diorite_stairs: MAKEID.polished_diorite_stairs,
    polished_granite_stairs: MAKEID.polished_granite_stairs,
    porkchop: MAKEID.porkchop,
    portal: MAKEID.portal,
    potato: MAKEID.potato,
    potatoes: MAKEID.potatoes,
    potion: MAKEID.potion,
    powered_comparator: MAKEID.powered_comparator,
    powered_rail: MAKEID.powered_rail,
    powered_repeater: MAKEID.powered_repeater,
    prismarine: MAKEID.prismarine,
    prismarine_bricks_stairs: MAKEID.prismarine_bricks_stairs,
    prismarine_crystals: MAKEID.prismarine_crystals,
    prismarine_shard: MAKEID.prismarine_shard,
    prismarine_stairs: MAKEID.prismarine_stairs,
    pufferfish: MAKEID.pufferfish,
    pumpkin: MAKEID.pumpkin,
    pumpkin_pie: MAKEID.pumpkin_pie,
    pumpkin_seeds: MAKEID.pumpkin_seeds,
    pumpkin_stem: MAKEID.pumpkin_stem,
    purple_glazed_terracotta: MAKEID.purple_glazed_terracotta,
    purpur_block: MAKEID.purpur_block,
    purpur_stairs: MAKEID.purpur_stairs,
    quartz: MAKEID.quartz,
    quartz_block: MAKEID.quartz_block,
    quartz_ore: MAKEID.quartz_ore,
    quartz_stairs: MAKEID.quartz_stairs,
    rabbit: MAKEID.rabbit,
    rabbit_foot: MAKEID.rabbit_foot,
    rabbit_hide: MAKEID.rabbit_hide,
    rabbit_stew: MAKEID.rabbit_stew,
    rail: MAKEID.rail,
    record_11: MAKEID.record_11,
    record_13: MAKEID.record_13,
    record_blocks: MAKEID.record_blocks,
    record_cat: MAKEID.record_cat,
    record_chirp: MAKEID.record_chirp,
    record_far: MAKEID.record_far,
    record_mall: MAKEID.record_mall,
    record_mellohi: MAKEID.record_mellohi,
    record_stal: MAKEID.record_stal,
    record_strad: MAKEID.record_strad,
    record_wait: MAKEID.record_wait,
    record_ward: MAKEID.record_ward,
    red_flower: MAKEID.red_flower,
    red_glazed_terracotta: MAKEID.red_glazed_terracotta,
    red_mushroom: MAKEID.red_mushroom,
    red_mushroom_block: MAKEID.red_mushroom_block,
    red_nether_brick: MAKEID.red_nether_brick,
    red_nether_brick_stairs: MAKEID.red_nether_brick_stairs,
    red_sandstone: MAKEID.red_sandstone,
    red_sandstone_stairs: MAKEID.red_sandstone_stairs,
    redstone: MAKEID.redstone,
    redstone_block: MAKEID.redstone_block,
    redstone_lamp: MAKEID.redstone_lamp,
    redstone_ore: MAKEID.redstone_ore,
    redstone_torch: MAKEID.redstone_torch,
    redstone_wire: MAKEID.redstone_wire,
    reeds: MAKEID.reeds,
    repeater: MAKEID.repeater,
    rotten_flesh: MAKEID.rotten_flesh,
    saddle: MAKEID.saddle,
    salmon: MAKEID.salmon,
    sand: MAKEID.sand,
    sandstone: MAKEID.sandstone,
    sandstone_stairs: MAKEID.sandstone_stairs,
    sapling: MAKEID.sapling,
    scaffolding: MAKEID.scaffolding,
    sea_pickle: MAKEID.sea_pickle,
    seagrass: MAKEID.seagrass,
    sealantern: MAKEID.sealantern,
    shears: MAKEID.shears,
    sheep: MAKEID.sheep,
    shield: MAKEID.shield,
    shulker_box: MAKEID.shulker_box,
    shulker_shell: MAKEID.shulker_shell,
    sign: MAKEID.sign,
    silver_glazed_terracotta: MAKEID.silver_glazed_terracotta,
    skeleton: MAKEID.skeleton,
    skull: MAKEID.skull,
    slime: MAKEID.slime,
    slime_ball: MAKEID.slime_ball,
    smithing_table: MAKEID.smithing_table,
    smoker: MAKEID.smoker,
    smooth_quartz_stairs: MAKEID.smooth_quartz_stairs,
    smooth_red_sandstone_stairs: MAKEID.smooth_red_sandstone_stairs,
    smooth_sandstone_stairs: MAKEID.smooth_sandstone_stairs,
    smooth_stone: MAKEID.smooth_stone,
    snow: MAKEID.snow,
    snow_layer: MAKEID.snow_layer,
    snowball: MAKEID.snowball,
    soul_sand: MAKEID.soul_sand,
    spawn_egg: MAKEID.spawn_egg,
    speckled_melon: MAKEID.speckled_melon,
    spider: MAKEID.spider,
    spider_eye: MAKEID.spider_eye,
    splash_potion: MAKEID.splash_potion,
    sponge: MAKEID.sponge,
    spruce_button: MAKEID.spruce_button,
    spruce_door: MAKEID.spruce_door,
    spruce_fence_gate: MAKEID.spruce_fence_gate,
    spruce_pressure_plate: MAKEID.spruce_pressure_plate,
    spruce_sign: MAKEID.spruce_sign,
    spruce_stairs: MAKEID.spruce_stairs,
    spruce_standing_sign: MAKEID.spruce_standing_sign,
    spruce_trapdoor: MAKEID.spruce_trapdoor,
    spruce_wall_sign: MAKEID.spruce_wall_sign,
    squid: MAKEID.squid,
    stained_glass: MAKEID.stained_glass,
    stained_glass_pane: MAKEID.stained_glass_pane,
    stained_hardened_clay: MAKEID.stained_hardened_clay,
    standing_banner: MAKEID.standing_banner,
    standing_sign: MAKEID.standing_sign,
    stick: MAKEID.stick,
    sticky_piston: MAKEID.sticky_piston,
    stickyPistonArmCollision: MAKEID.stickyPistonArmCollision,
    stone: MAKEID.stone,
    stone_axe: MAKEID.stone_axe,
    stone_brick_stairs: MAKEID.stone_brick_stairs,
    stone_button: MAKEID.stone_button,
    stone_hoe: MAKEID.stone_hoe,
    stone_pickaxe: MAKEID.stone_pickaxe,
    stone_pressure_plate: MAKEID.stone_pressure_plate,
    stone_shovel: MAKEID.stone_shovel,
    stone_slab: MAKEID.stone_slab,
    stone_slab_2: MAKEID.stone_slab_2,
    stone_slab_3: MAKEID.stone_slab_3,
    stone_slab_4: MAKEID.stone_slab_4,
    stone_stairs: MAKEID.stone_stairs,
    stone_sword: MAKEID.stone_sword,
    stonebrick: MAKEID.stonebrick,
    stonecutter_block: MAKEID.stonecutter_block,
    stray: MAKEID.stray,
    strider: MAKEID.strider,
    string: MAKEID.string,
    stripped_acacia_log: MAKEID.stripped_acacia_log,
    stripped_birch_log: MAKEID.stripped_birch_log,
    stripped_dark_oak_log: MAKEID.stripped_dark_oak_log,
    stripped_jungle_log: MAKEID.stripped_jungle_log,
    stripped_oak_log: MAKEID.stripped_oak_log,
    stripped_spruce_log: MAKEID.stripped_spruce_log,
    structure_block: MAKEID.structure_block,
    structure_void: MAKEID.structure_void,
    sugar: MAKEID.sugar,
    suspicious_stew: MAKEID.suspicious_stew,
    sweet_berries: MAKEID.sweet_berries,
    tallgrass: MAKEID.tallgrass,
    tnt: MAKEID.tnt,
    tnt_minecart: MAKEID.tnt_minecart,
    torch: MAKEID.torch,
    totem: MAKEID.totem,
    trapdoor: MAKEID.trapdoor,
    trapped_chest: MAKEID.trapped_chest,
    trident: MAKEID.trident,
    tripWire: MAKEID.tripWire,
    tripwire_hook: MAKEID.tripwire_hook,
    tropicalfish: MAKEID.tropicalfish,
    turtle: MAKEID.turtle,
    turtle_helmet: MAKEID.turtle_helmet,
    turtle_shell_piece: MAKEID.turtle_shell_piece,
    undefined: MAKEID.undefined,
    undyed_shulker_box: MAKEID.undyed_shulker_box,
    unlit_redstone_torch: MAKEID.unlit_redstone_torch,
    unpowered_comparator: MAKEID.unpowered_comparator,
    unpowered_repeater: MAKEID.unpowered_repeater,
    villager: MAKEID.villager,
    villager_v2: MAKEID.villager_v2,
    vine: MAKEID.vine,
    wall_banner: MAKEID.wall_banner,
    wall_sign: MAKEID.wall_sign,
    wandering_trader: MAKEID.wandering_trader,
    water: MAKEID.water,
    waterlily: MAKEID.waterlily,
    web: MAKEID.web,
    wheat: MAKEID.wheat,
    wheat_seeds: MAKEID.wheat_seeds,
    white_glazed_terracotta: MAKEID.white_glazed_terracotta,
    witch: MAKEID.witch,
    wither_rose: MAKEID.wither_rose,
    wither_skeleton: MAKEID.wither_skeleton,
    wither_skull_dangerous: MAKEID.wither_skull_dangerous,
    wolf: MAKEID.wolf,
    wood: MAKEID.wood,
    wooden_axe: MAKEID.wooden_axe,
    wooden_button: MAKEID.wooden_button,
    wooden_door: MAKEID.wooden_door,
    wooden_hoe: MAKEID.wooden_hoe,
    wooden_pickaxe: MAKEID.wooden_pickaxe,
    wooden_pressure_plate: MAKEID.wooden_pressure_plate,
    wooden_shovel: MAKEID.wooden_shovel,
    wooden_slab: MAKEID.wooden_slab,
    wooden_sword: MAKEID.wooden_sword,
    wool: MAKEID.wool,
    writable_book: MAKEID.writable_book,
    written_book: MAKEID.written_book,
    xp_bottle: MAKEID.xp_bottle,
    xp_orb: MAKEID.xp_orb,
    yellow_flower: MAKEID.yellow_flower,
    yellow_glazed_terracotta: MAKEID.yellow_glazed_terracotta,
    zombie: MAKEID.zombie,
    zombie_pigman: MAKEID.zombie_pigman,
    zomebie_villager_v2: MAKEID.zomebie_villager_v2,
};
export default ID;
