// @mcbe/farticker/index.ts
import { Box } from "krgeometry";
import { command } from "@mcbe/system_server";
import Identifier from "@mcbe/identifier";
import ID from "@mcbe/identifier/id";
import { setblockWithoutError } from "@mcbe/command";

const ids:number[] = [];
for (let i=0;i<10;i++)
{
    ids[i] = i;
}

const waiting:Processing[] = [];
const processing = new Set<Processing>();
let processingTimeout:NodeJS.Timeout|null = null;

class Processing
{
    private index = 0;
    private processing = false;

    constructor(
        public id:number, 
        private readonly blocks:[Identifier, number?][],
        public readonly pos:VectorXYZ, 
        public readonly area:Box, 
        public readonly cb:(last:Identifier, data?:number)=>(Promise<boolean>|boolean),
        private readonly resolve:()=>void)
    {
        command(`tickingarea add ${area.x1} ${area.y1} ${area.z1} ${area.x2} ${area.y2} ${area.z2} TR${id}`);
    }

    async process():Promise<void>
    {
        try
        {
            const pos = this.pos;
            this.index = (this.index + 1) % this.blocks.length;
    
            const block = this.blocks[this.index];
            if (this.processing) return;
    
            this.processing = true;
            const res = await setblockWithoutError(pos, block[0], block[1]);
            if (!res)
            {
                this.processing = false;
                return;
            }
            try
            {
                const res = await this.cb(block[0], block[1]);
                if (!res)
                {
                    this.processing = false;
                    return;
                }
            }
            catch(err)
            {
                console.error(err);
            }
            command(`tickingarea remove TR${this.id}`);
            const w = waiting.pop();
            if (w)
            {
                w.id = this.id;
                processing.add(w);
            }
            else
            {
                processing.delete(this);
                ids.push(this.id);
            }
            this.processing = false;
            this.resolve();
        }
        catch(err)
        {
            console.error(err);
        }
    }
}

function process():void
{
    if (processing.size === 0)
    {
        if (processingTimeout)
        {
            clearInterval(processingTimeout);
            processingTimeout = null;
        }
    }
    for (const p of processing)
    {
        p.process();
    }
}

function addProcessing(pos:VectorXYZ, area:Box, list:[Identifier, number?][], cb:(last:Identifier, data?:number)=>Promise<boolean>|boolean):Promise<void>
{
    return new Promise<void>(resolve => {
        const id = ids.pop();
        if (id === undefined)
        {
            waiting.push(new Processing(-1, list, pos, area, cb, resolve));
        }
        else
        {
            processing.add(new Processing(id, list, pos, area, cb, resolve));
            if (!processingTimeout)
            {
                processingTimeout = setInterval(process, 1000);
            }
        }
    });
}

/**
 * Add ticking area to area and callback
 * @param pos position to test, It will overwrite stone or dirt to this position
 * @param area position to add ticking area
 * @param cb callback when it loaded, it will be called again if it returns false
 */
export function setFarTicker(pos:VectorXYZ, area:Box, cb:()=>Promise<boolean>|boolean):Promise<void>
{
    return addProcessing(pos, area, [[ID.stone], [ID.dirt]], cb);
}

/**
 * set block to position
 */
export async function setFarBlock(pos:VectorXYZ, block:Identifier, data?:number):Promise<void>
{
    const res = await setblockWithoutError(pos, block, data);
    if (res) return;

    const second = block === ID.stone ? ID.dirt : ID.stone;
    await addProcessing(pos, Box.fromVector(pos, pos), [[block, data], [second]], last => {
        if (last === block) return true;
        return setblockWithoutError(pos, block, data);
    });
}
