// @mcbe/ruacore_client/index.ts

import { createBroadcast, evitem, createListener } from "@mcbe/event";
import Event from "krevent";
import { system } from "@mcbe/system_client";

const ruacore = {
    initialize:new Event<()=>void>(),
    shutdown:new Event<()=>void>(),
    update:new Event<()=>void>(),
    broadcast: createBroadcast({
        DisplayChat: evitem<IDisplayChatParameters>(SendToMinecraftClient.DisplayChat),
        LoadUI: evitem<ILoadUIParameters>(SendToMinecraftClient.LoadUI),
        ScriptLoggerConfig: evitem<IScriptLoggerConfigParameters>(SendToMinecraftClient.ScriptLoggerConfig),
        SendUIEvent: evitem<ISendUIEventParameters>(SendToMinecraftClient.SendUIEvent),
        SpawnParticleAttachedEntity: evitem<ISpawnParticleAttachedEntityParameters>(SendToMinecraftClient.SpawnParticleAttachedEntity),
        SpawnParticleInWorld: evitem<ISpawnParticleInWorldParameters>(SendToMinecraftClient.SpawnParticleInWorld),
        UnloadUI: evitem<IUnloadUIParameters>(SendToMinecraftClient.UnloadUI),
    }),
    listen: createListener({
        ClientEnteredWorld: evitem<IClientEnteredWorldEventData>(ReceiveFromMinecraftClient.ClientEnteredWorld),
        HitResultChanged: evitem<IHitResultChangedEventData>(ReceiveFromMinecraftClient.HitResultChanged),
        HitResultContinuous: evitem<IHitResultContinuousEventData>(ReceiveFromMinecraftClient.HitResultContinuous),
        PickHitResultChanged: evitem<IPickHitResultChangedEventData>(ReceiveFromMinecraftClient.PickHitResultChanged),
        PickHitResultContinuous: evitem<IPickHitResultContinuousEventData>(ReceiveFromMinecraftClient.PickHitResultContinuous),
        UIEvent: evitem<string>(ReceiveFromMinecraftClient.UIEvent),
    }),
};
Object.freeze(ruacore);

system.initialize = ()=>ruacore.initialize.fire();
system.shutdown = ()=>ruacore.shutdown.fire();
system.update = ()=>ruacore.update.fire();
Object.freeze(system);

export default ruacore;
